rpcd/playbooks/roles/rpc_maas/library/raxmon.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python
+
+# Copyright 2017, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ConfigParser
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def _get_agent_tokens(conn, entity):
+    agent_tokens = []
+    for a in conn.list_agent_tokens():
+        if a.label == entity:
+            agent_tokens.append(a)
+    return agent_tokens
+
+
+def _get_conn(get_driver, Provider, raxmon_cfg):
+    cfg = ConfigParser.RawConfigParser()
+    cfg.read(raxmon_cfg)
+    driver = get_driver(Provider.RACKSPACE)
+    conn = None
+    try:
+        user = cfg.get('credentials', 'username')
+        api_key = cfg.get('credentials', 'api_key')
+        conn = driver(user, api_key)
+    except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
+        url = cfg.get('api', 'url')
+        token = cfg.get('api', 'token')
+        conn = driver(None, None, ex_force_base_url=url,
+                      ex_force_auth_token=token)
+    return conn
+
+
+def _get_entities(conn, entity):
+    entities = []
+    for e in conn.list_entities():
+        if e.label == entity:
+            entities.append(e)
+    return entities
+
+
+def assign_agent_to_entity(module, conn, entity):
+    entities = _get_entities(conn, entity)
+    entities_count = len(entities)
+    if entities_count == 0:
+        msg = "Zero entities with the label %s exist. Entities should be " \
+              "created as part of the hardware provisioning process, if " \
+              "missing, please consult the internal documentation for " \
+              "associating one with the device." % (entity)
+        module.fail_json(msg=msg)
+    elif entities_count == 1:
+        conn.update_entity(entities[0], {'agent_id': entity})
+        module.exit_json(
+            changed=True
```

Comment:
Hmm, similar to https://github.com/rcbops/rpc-openstack/blob/9fee12fbb659a5291069b1a39a2787d478e270f3/rpcd/playbooks/roles/rpc_maas/tasks/agent_install_remote.yml#L51-L56 this will always update the entity and therefore it should always return changed.  However, I'll update this to check if entities[0].label == entities[0].agent_id and if so exit with changed=False.  Otherwise, it'll update the entity and exit with changed=True.